packages/geoip-location/src/GeoIpLocator.ts

```diff
@@ -0,0 +1,109 @@
+    private readonly dbCheckInterval: number
+    private readonly dbCheckErrorInterval: number
+    private readonly mirrorUrl?: string
+    private reader?: Reader<CityResponse>
+    private dbCheckTimeout?: LongTimeout.Timeout
+
+    constructor(
+        geoIpDatabaseFolder: string,
+        dbCheckInterval = DEFAULT_DB_CHECK_INTERVAL,
+        dbCheckErrorInterval = DEFAULT_DB_CHECK_ERROR_INTERVAL,
+        mirrorUrl?: string
+    ) {
+        this.abortController = new AbortController()
+        this.dbCheckInterval = dbCheckInterval
+        this.dbCheckErrorInterval = dbCheckErrorInterval
+        if (!geoIpDatabaseFolder.endsWith('/')) {
+            geoIpDatabaseFolder += '/'
+        }
+        this.geoIpDatabaseFolder = filePathToNodeFormat(geoIpDatabaseFolder)
+        this.mirrorUrl = mirrorUrl
+    }
+
+    private checkDatabase: () => Promise<void> = async () => {
+        if (this.reader === undefined) {
+            // if we do not have a reader, create a new one in any case
+            this.reader = await downloadGeoIpDatabase(this.geoIpDatabaseFolder, true, this.abortController.signal, this.mirrorUrl)
+        } else {
+            // if we already have a reader, create a new one only if db has changed
+            const newReader = await downloadGeoIpDatabase(this.geoIpDatabaseFolder, false, this.abortController.signal, this.mirrorUrl)
+            if (newReader !== undefined) {
+                this.reader = newReader
+            }
+        }
+    }
+
+    private scheduleCheck: (timeout: number) => void = async (timeout: number) => {
+        if (this.abortController.signal.aborted) {
+            return
+        }
+        this.dbCheckTimeout = LongTimeout.setTimeout(async () => {
+            try {
+                await this.checkDatabase()
+                this.scheduleCheck(this.dbCheckInterval)
+            } catch (err) {
+                logger.warn('GeoIpLocator: GeoIP database check failed', { err })
+                this.scheduleCheck(this.dbCheckErrorInterval)
+            } 
+        }, timeout)
+    }
+
+    async start(): Promise<void> {
+        if (this.dbCheckTimeout !== undefined) {
+            return
+        }
+
+        await this.checkDatabase()
+        this.scheduleCheck(this.dbCheckInterval)
+    }
+
+    stop(): void {
+        if (this.dbCheckTimeout !== undefined) {
+            LongTimeout.clearTimeout(this.dbCheckTimeout)
+        }
+        this.abortController.abort()
+    }
+
+    lookup(ip: string): GeoIpLookupResult | undefined {
+        if (this.reader === undefined) {
+            logger.warn('GeoIpLocator: lookup called before database is ready (maybe start() was not called?')
+            return undefined
+        }
+
+        // If ip is undefined of null, the library will crash
+        // this might happen despite the ts typings because the ip address 
+        // comes from the ws server socket and is not under our control
+
+        if (!(ip as unknown)) {
```

Comment:
Is there any case where this would happen in the real world?

Reply:
It did happen while testing, and that's why I added the check. 

In all, this is a timer-run system that is run very rarely, so we should make it as difficult to crash as possible, as crashes have a high probability of staying unnoticed ("It has been running fine for 20 days, so probably it will keep that way")